### PR TITLE
PSC monitoring MVP

### PR DESCRIPTION
**Pull request by cbiffle**
This adds very basic monitoring of the presence and power-good lines from the PSUs at the PSC. It can manage remove/insertion of power supplies, delay restarts to ensure we don't cycle things too quickly, and respond to faults by resetting things. So far data is only logged to a ringbuf.

### CODE REVIEW

File: drv/psc-seq-server/src/main.rs
Comment (hawkw) [inline drv/psc-seq-server/src/main.rs:509]: _very_ nitpicky, to the extent that i wasn't sure if i should even mention this: elsewhere, you've put the comment describing what case the states represent _above_ the match arm, but here it's under it:
```suggestion
            // The PSU appears to have pulled the OK signal into the "not  OK" 
            // state to indicate an internal fault!
            (PsuState::Present(PresentState::On), _, Status::NotGood) => {
```

File: drv/psc-seq-server/src/main.rs
Comment (hawkw) [inline drv/psc-seq-server/src/main.rs:546]: take it or leave it: I kind of feel like this might be a little bit easier to follow if we pulled these out into separate match arms? not a huge deal though, it's a marginal improvement at best.
```suggestion
            // Do it all again then, I guess.
            (
                PsuState::Present(PresentState::Faulted { turn_on_deadline }),
                _,
                Status::NotGood,
            ) if turn_on_deadline <= now => {
                let turn_on_deadline =
                    now.wrapping_add(FAULT_OFF_MS);
                self.state =
                    PsuState::Present(PresentState::Faulted {
                        turn_on_deadline,
                    });
                Some(ActionRequired::NoteStickyFault)
            },

            // The PSU has indicated that it is once again good!
            (
                PsuState::Present(PresentState::Faulted { turn_on_deadline }),
                _,
                Status::Good,
            ) if turn_on_deadline <= now => {
                self.state = PsuState::Present(PresentState::On);
                Some(ActionRequired::EnableMe)
            }

            // Remain in this state.
            _ => None
```

File: drv/psc-seq-server/src/main.rs
Comment (hawkw) [inline drv/psc-seq-server/src/main.rs:597]: nit, take it or leave it: I might consider defining these enums _before_ the logic that operates on them, rather than after --- the documentation here is valuable context when reading the implementations of those methods, so it might be nice if a top-to-bottom reader encountered it first? not a big deal though.

File: drv/psc-seq-server/src/main.rs
Comment (hawkw) [inline drv/psc-seq-server/src/main.rs:447]: i kind of wonder if an ascii graphic flowchart depicting these state transitions might be useful?

File: app/psc/base.toml
Comment (cbiffle) [reply app/psc/base.toml:193]: Yeah I don't understand the TOML stuff well enough to do that. I can check with Matt.

File: drv/psc-seq-server/src/main.rs
Comment (labbott) [inline drv/psc-seq-server/src/main.rs:36]: The "this task will start anew" seems to be in conflict with the sentence below "This task is not intended to restart under normal operation". I _think_ this paragraph is describing cold boot behavior vs below is runtime?

File: app/psc/base.toml
Comment (hawkw) [reply app/psc/base.toml:193]: It's not a big deal either way, just occurred to me that we could save some space on the downrev boards by avoiding the separate task. 🤷‍♀️

File: drv/psc-seq-server/src/main.rs
Comment (cbiffle) [reply drv/psc-seq-server/src/main.rs]: Fair point, fixed!

File: drv/psc-seq-server/src/main.rs
Comment (cbiffle) [reply drv/psc-seq-server/src/main.rs:509]: Interesting, these two classes of comments are very different things in my head, but not in a way I can easily explain or make actionable for future maintainers.

Basically, there's one class of comment explaining the _pattern,_ and one class explaining what to do after the pattern matches. e.g. the ones above the match arms are generally explaining why a field is removed or wildcard, the ones below talk strategy.

But I get that it just looks inconsistent, so I can make them all the same.

File: drv/psc-seq-server/src/main.rs
Comment (cbiffle) [reply drv/psc-seq-server/src/main.rs:546]: Hmmmm I'm not sure I feel the same on this.

I do think _all_ of these match arms were easier to read before rustfmt made them v
e
r
t
i
c
a
l
:\

File: drv/psc-seq-server/src/main.rs
Comment (cbiffle) [reply drv/psc-seq-server/src/main.rs:597]: Done!

File: drv/psc-seq-server/src/main.rs
Comment (hawkw) [reply drv/psc-seq-server/src/main.rs:509]: Ah, interesting, I see what you mean! To me, the comments on the patterns seemed to describe, in a general sense, "what this match arm represents", so this felt more like that. But, it's up to you --- do whatever you think is best!

File: drv/psc-seq-server/src/main.rs
Comment (cbiffle) [reply drv/psc-seq-server/src/main.rs:36]: Yeah, basically -- the PSC getting plugged in means _power is applied._ I have attempted to clarify this.

**Comment by cbiffle** [comment]
Fixes pushed atop the original commit, PTAL

File: drv/psc-seq-server/src/main.rs
Comment (hawkw) [reply drv/psc-seq-server/src/main.rs:600]: My thought process was that to me, when the match arm captured the entire condition that represented a particular state transition, rather than having additional branching logic inside it, you could view the match as a mapping from "all the inputs" to "all the outputs", which felt maybe slightly more comprehensible to _my brain_ in particular...but I think this is a matter of taste, and I'm fine with the way you have it.

File: drv/psc-seq-server/src/main.rs
Comment (hawkw) [inline drv/psc-seq-server/src/main.rs:443]: this is _very_ nitpicky, but it feels like the naming of this variable seems to be a bit in conflict over whether it represents a _fact_ about the state of the world ("is the PSU `present`?') or a _policy_ about whether or not to perform a particular action ("should we `attempt_snapshot`?"). 

it feels like there is a potential for future issues if we were to later change `step` to decide whether or not to attempt a snapshot based on more factors than _just_ whether the PSU is present, without updating this code. if we were to ever return `attempt_snapshot: false` for a present PSU for some reason, the ringbuffer would incorrectly indicate that the PSU is not present.

just a thought.

**Review by hawkw — APPROVED** [review]
no blockers from me!

File: drv/psc-seq-server/src/main.rs
Comment (cbiffle) [reply drv/psc-seq-server/src/main.rs:443]: I think it'll be clearer once I finish and merge the PMBus snapshot code, but I may rename it based on your suggestion at that point. The two names are a result of two different use cases:
- It was going to be named `attempt_snapshot` internally to reflect the implementation intent, but then
- I recognized that that name wasn't very useful in a ringbuf and came up with a different one for logging there. (This one's the user-visible one.)

**Pull request merged**
